src/webserver/webcache/WebLruCache.java

```diff
@@ -0,0 +1,195 @@
+   * 
+   * @param maxCapacity the max capacity of this cache.
+   */
+  public WebLruCache(int maxCapacity) {
+    this.maxCapacity = maxCapacity;
+    this.lookup = new HashMap<>(maxCapacity);
+
+    initializeCleanupThread();
+  }
+
+  /**
+   * Puts an object in the cache.
+   * <p>
+   * The object will be stored under the specified id, for later retrieval. If
+   * duplicate ids are provide, the old id and object are overwritten.
+   * <p>
+   * The cached object will only live for the specified amount of time, after
+   * which it expires and is removed from the cache.
+   * 
+   * @param item          the item to cache.
+   * @param id            the id associated with this item.
+   * @param secondsToLive the amount of seconds this item has to live.
+   */
+  public synchronized void putCache(String item, String id, int secondsToLive) {
+    synchronized (lookup) {
+      if (this.lookup.size() >= this.maxCapacity) {
+        // Remove the tail from the lookup index in O(n) worst time
+        for (String key : this.lookup.keySet()) {
+          if (this.lookup.get(key).equals(tail)) {
+            this.lookup.remove(key);
+            break;
+          }
+        }
+  
+        // Remove other references
+        if (tail.prev != null) {
+          tail.prev.next = null;
+        }
+  
+        tail = tail.prev;
+      }
+  
+      // Create and insert the new node into the proper areas
+      Node<TimedObject<String>> newNode = new Node<TimedObject<String>>(new TimedObject<String>(item, secondsToLive),
+          null, head);
+      this.head = newNode;
+      if (this.lookup.size() == 0) {
+        this.tail = newNode;
+      }
+  
+      this.lookup.put(id, newNode);
+    }
+  }
+
+  /**
+   * Updates a cached item.
+   * <p>
+   * The object stored under the existing id will be swapped with the new item. If
+   * the id does not exist, nothing happens.
+   * <p>
+   * This method is similar to calling {@link #putCache(String, String, int)} with
+   * a replacement object, but using this method indicates an update. This cached
+   * object will also retain the existing amount of seconds to live from the old
+   * object.
+   * 
+   * @param id      the id to update.
+   * @param newItem the new item to replace the old item.
+   */
+  public synchronized void updateCache(String id, String newItem) {
+    synchronized (this.lookup) {
+      if (checkCache(id)) {
+        this.lookup.get(id).data.updateObject(newItem);
+      }
+    }
+  }
+
+  /**
+   * Attempts to retrieve a cached object using a specified key.
+   * <p>
+   * If the cached object and key does not exist, {@code null} will be
+   * returned.
+   * 
+   * @param id the id of the cached object to fetch.
+   * @return the cached object, or {@code null}.
+   */
+  public synchronized String getCachedObject(String id) {
+    synchronized (this.lookup) {
+        return this.lookup.get(id).data.getObject();
+    }
+  }
```

Comment:
Forgot to bump the node to the head, thus making it a FIFO cache and not an LRU cache. Also, more synchronized pain.